texastribune/donations
# Add donation creator func

#### What's this PR do?
Adds a function that takes some vars and builds a basic recurring donation (membership level) out of them.

#### Why are we doing this? How does it help us?
This is the starter piece for an eventual larger tool in the portal to allow users to update their giving amount on a recurring donation. That larger tool will involve deleting the current recurring donation (a piece we already have) and creating a new recurring donation (this piece). Until we're there though, this piece is crucial to allow us to complete that amount change request when it comes in from a member (through the membership team).

#### How should this be manually tested?
* Ensure you have these env vars in your env-docker file
    * PUBLISHABLE_KEY
    * SECRET_KEY
    * STRIPE_WEBHOOK_SECRET
    * STRIPE_PRODUCT_SUSTAINING
* Visit the customers page of our [stripe dashboard in test mode](https://dashboard.stripe.com/test/customers)
* Click on a customer
* Find their customer id on the left in the drop-down titled "Details" and copy it
    * <img width="408" alt="Screen Shot 2023-10-24 at 4 16 31 PM" src="https://github.com/texastribune/donations/assets/88053677/c435a90e-862d-4720-9d9e-d4605b820bba">
* Locally, run `make shell`
* From here do something like...
    ```
    from server.util import donation_adder
    d = donation_adder(
            customer="{customer_id copied above}",
            amount=11,
            pay_fees=True,
            interval="month",
            year=2023,
            month=10,
            day=26)
    ```
* Visit the stripe page again and refresh to see if the subscription came through
* Visit our salesforce page for testing
* Use the search bar to look up the email address of the customer you added the donation for
    * <img width="1134" alt="Screen Shot 2023-10-24 at 4 26 24 PM" src="https://github.com/texastribune/donations/assets/88053677/28093449-07c8-40af-8553-f043c47741c6">
* On the customer page, find the "Recurring Donations" tab and click on the "Date Established" column heading to see all the rdos most recent to oldest
    * <img width="1124" alt="Screen Shot 2023-10-24 at 4 27 39 PM" src="https://github.com/texastribune/donations/assets/88053677/3643bcd5-9568-44b5-ac89-9b08da42f202">
* You should see the 10/26 donation towards the top
    * <img width="1113" alt="Screen Shot 2023-10-24 at 4 30 03 PM" src="https://github.com/texastribune/donations/assets/88053677/96b70212-748e-4cb0-b559-eaceb13e9a7d">
* Click through and make sure all looks right
    * amount
    * date
    * opportunities
    * existent card info

#### How should this change be communicated to end users?
I'll let membership and Anna know we have a tool to add subs until the ump feature is up and running.

#### Are there any smells or added technical debt to note?
This is definitely a stop gap tool since we aren't immediately working on an UMP feature to do this. While this is very in the weeds of this project (and admittedly hacky), I'll reutilize most of this code to build the aforementioned ump feature (and to migrate existing salesforce donations to stripe), and once that's working, we can remove this func.

#### What are the relevant tickets?
https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwS1XPty68eK4Ett/recGrExlnN6l32LS9?blocks=hide

#### Have you done the following, if applicable:
***(optional: add explanation between parentheses)***


#### TODOs / next steps:
